### PR TITLE
BBS-83 - Reservation list

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

File: modules/aleph/includes/aleph.reservation.inc [hunk 1]
```diff
@@ -17,8 +23,76 @@ function aleph_reservation_options_submit() {
...
-  return array();
+/**
+ * Provider list callback.
+ *
+ * @param object $account
+ *    User to list reservations for.
+ * @param string|null $requested_type
+ *    Type of reservations to list. NULL for all.
+ *
+ * @return array|mixed
```
Comment (xendk) [inline modules/aleph/includes/aleph.reservation.inc]: Doesn't it always return an array?

File: modules/aleph/includes/aleph.reservation.inc [hunk 2]
```diff
@@ -37,7 +111,28 @@ function aleph_reservation_reservation_deletion_enabled() {
...
+ * Return a branch name for a given branch id.
+ *
+ * @param string $branch_id
+ *   The identifier for a given pick-up branch (e.g. BBAAA).
+ *
+ * @return string
+ *   The human name for the id or NULL if one do not exist.
+ */
+function aleph_reservation_branch_name($branch_id) {
+  if (strlen($branch_id) > 5) {
```
Comment (xendk) [inline modules/aleph/includes/aleph.reservation.inc]: Why this? Is the branch_id in fact the human-readable name in some cases? Then it needs to be documented.

File: modules/aleph/includes/aleph.reservation.inc [hunk 3]
```diff
@@ -37,7 +111,28 @@ function aleph_reservation_reservation_deletion_enabled() {
...
+    return $branch_id;
+  }
+
+  $branches = array(
```
Comment (xendk) [inline modules/aleph/includes/aleph.reservation.inc:133]: I do hope this is just a placeholder until some branch fetching code is written.

In which case it's missing a `TODO` comment.

File: modules/aleph/src/Aleph/AlephRequest.php [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+ *
+ * Entity with setters/getters for Z37 (REQUESTS) in Aleph.
+ *
+ * The Z37 contains requests information.
+ * The Z37 record functions both for requests for
+ * loaned items (waiting requests), and for requests for delivery of available
+ * items. It is not necessarily expected that a library will convert requests
+ * when installing the ALEPH system. In many cases, the library will inform
+ * patrons that there is a system switchover, that outstanding requests are
+ * not retained, and must be re-entered after switchover.
```
Comment (xendk) [inline modules/aleph/src/Aleph/AlephRequest.php]: How a library handles switchover from one system to another is irrelevant for this code, this comment only confuses.

File: modules/aleph/src/Aleph/AlephRequest.php [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+   * Hold request status.
+   *
+   * A = Active (newly opened, not yet processed).
+   *
+   * S = Hold Shelf - Request has been trapped, patron has been notified that
+   * the requested material is ready to be picked up.
+   *
+   * W = Waiting. System generated when the "print call slip" service
+   * (b- cir-12) was run, and the system detected that there was not an
+   * available copy
```
Comment (xendk) [inline modules/aleph/src/Aleph/AlephRequest.php]: Too much description about what's goes on behind the scenes.

File: modules/aleph/src/Aleph/AlephRequest.php [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+   *
+   * @var string
+   */
+  protected $id;
+
+  /**
+   * Z37-PICKUP-LOCATION.
+   * User’s pickup sub-library.
+   *
+   * The default is the user's home sub-library.
```
Comment (xendk) [inline modules/aleph/src/Aleph/AlephRequest.php]: That's describing the backend again. All we need to know is that it's the pickup location.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 7]
```diff
@@ -119,6 +124,46 @@ public function getDebts() {
...
+  public function getReservations() {
+    $reservations = array();
+    $hold_requests = $this->client->getReservations($this->getPatron())->xpath('hold-requests/institution/hold-request');
+    foreach ($hold_requests as $hold_request) {
+      if ((string) $hold_request->xpath('z37/z37-request-type')[0] === 'Hold Request') {
+        $reservation = new AlephReservation();
+        $request = new AlephRequest();
+        $material = new AlephMaterial();
+
+        // Create the request object.
```
Comment (xendk) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Well, obviously.

**Review by xendk — CHANGES_REQUESTED** [review]

File: modules/aleph/includes/aleph.reservation.inc (hunk 3 above)
Comment (tripox) [reply modules/aleph/includes/aleph.reservation.inc:133]: Yup, todo added, @xendk. :)

File: modules/aleph/src/Aleph/AlephRequest.php [hunk 8]
```diff
@@ -0,0 +1,380 @@
+<?php
+
+namespace Drupal\aleph\Aleph;
+
+/**
+ * Class AlephRequest.
+ *
+ * Entity with setters/getters for Z37 (REQUESTS) in Aleph.
+ *
+ * The Z37 contains requests information.
```
Comment (xendk) [inline modules/aleph/src/Aleph/AlephRequest.php]: You gotta loose all these references to Z37. It's the job of your API to *hide* all the complexity of how things is working behind the scenes, so don't throw it all in the face of the API user.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 9]
```diff
@@ -119,6 +124,42 @@ public function getDebts() {
...
+        $request->setPickupLocation((string) $hold_request->xpath('z37/z37-pickup-location')[0]);
+        $request->setOpenDate((string) $hold_request->xpath('z37/z37-open-date')[0]);
+        $request->setEndRequestDate((string) $hold_request->xpath('z37/z37-end-request-date')[0]);
+        $request->setDocNumber((string) $hold_request->xpath('z37/z37-doc-number')[0]);
+        $request->setHoldDate((string) $hold_request->xpath('z37/z37-hold-date')[0]);
+        $request->setRequestNumber((string) $hold_request->xpath('z37/z37-request-number')[0]);
+        $request->setSequence(ltrim((string) $hold_request->xpath('z37/z37-sequence')[0], 0));
+
+        $material->setTitle((string) $hold_request->xpath('z13/z13-title')[0]);
+        $material->setId((string) $hold_request->xpath('z13/z13-doc-number')[0]);
```
Comment (xendk) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php:151]: All this could be a mapping array and a loop.

**Review by xendk — CHANGES_REQUESTED** [review]

**Review by xendk — APPROVED** [review]

**Pull request merged**